raster/src/main/scala/geotrellis/raster/io/geotiff/GeoTiffSegmentLayout.scala

```diff
@@ -122,14 +125,159 @@ case class GeoTiffSegmentLayout(totalCols: Int, totalRows: Int, tileLayout: Tile
   def intersectingSegments(bounds: GridBounds): Array[Int] = {
     val tc = tileLayout.tileCols
     val tr = tileLayout.tileRows
-    val ab = ArrayBuffer[Int]()
+    val ab = mutable.ArrayBuffer[Int]()
     for (layoutCol <- (bounds.colMin / tc) to (bounds.colMax / tc)) {
       for (layoutRow <- (bounds.rowMin / tr) to (bounds.rowMax / tr)) {
         ab += (layoutRow * tileLayout.layoutCols) + layoutCol
       }
     }
     ab.toArray
   }
+
+  /** Partition a list of pixel windows to localize required segment reads.
+    * Some segments may be required by more than one partition.
+    * Pixel windows outside of layout range will be filtered.
+    * Maximum partition size may be exceeded if any window size exceeds it.
+    * Windows will not be split to satisfy partition size limits.
+    *
+    * @param windows List of pixel windows from this layout
+    * @param maxPartitionSize Maximum pixel count for each partition
+    */
+  def partitionWindowsBySegments(
+    windows: Seq[GridBounds],
+    maxPartitionSize: Long
+  ): Array[Array[GridBounds]] = {
+    val partition = mutable.ArrayBuilder.make[GridBounds]
+    partition.sizeHintBounded(128, windows)
```

Comment:
Hm, why 128?

Reply:
Pretty safe guess I guess. If each window is the common case of 256x256 and a double 128 gets you up to 68MB.